prov/efa/src/rdm/efa_rdm_ope.c

```diff
@@ -449,68 +450,105 @@ size_t efa_rdm_txe_max_req_data_capacity(struct efa_rdm_ep *ep, struct efa_rdm_o
 }
 
 /**
- * @brief set the max_req_data_size field of a txe for multi-req
- *
- * Multi-REQ protocols send multiple REQ packets via one call to ibv_post_send() for efficiency.
- * Under such circumstance, it is better that the data size of mulitple REQ packets to be close.
+ * @brief prepare ope to send the give pkt type
  *
- * To achieve the closeness, the field of max_req_data_size was introduced to efa_rdm_ope,
- * and used to limit data size when construct REQ packet.
+ * For given packet type, calcuate how many packets are going to be
+ * sent, and how many data each packet will carry.
  *
- * This function set the max_req_data_size properly.
+ * For runting read rtm, also set the "runt_size" field.
  *
- *
- * @param[in]		ep			endpoint
- * @param[in,out]	txe	txe that has all information of
+ * @param[in]		ep		endpoint
+ * @param[in,out]	ope		txe that has all information of
  * 					a send operation
  * @param[in]		pkt_type	type of REQ packet
- *
+ * @param[in]		pkt_entry_cnt	number of packets to be sent
+ * @param[in]		pkt_entry_data_size_vec data size of each packet. This field
+ *                                              is used when pkt_entry_cnt > 1.
+ *                                              (for DATA, MEDIUM and RUNTREAD)
+ * @return
+ * On success, return 0
+ * If there is not enough available packet entry in TX packet pool, return -FI_EAGAIN
  */
-void efa_rdm_txe_set_max_req_data_size(struct efa_rdm_ep *ep, struct efa_rdm_ope *txe, int pkt_type)
+ssize_t efa_rdm_ope_prepare_to_post_send(struct efa_rdm_ope *ope,
+					 int pkt_type,
+					 int *pkt_entry_cnt,
+					 int *pkt_entry_data_size_vec)
 {
-	int max_req_data_capacity;
-	int mulreq_total_data_size;
-	int num_req;
-	int memory_alignment = 8;
-	size_t max_req_data_capacity_aligned;
+	struct efa_rdm_ep *ep;
+	size_t total_pkt_entry_data_size; /* total number of bytes send via packet entry's payload */
+	size_t single_pkt_entry_data_size;
+	size_t single_pkt_entry_max_data_size;
+	int i, memory_alignment = 8, remainder;
+	int available_tx_pkts;
 
-	assert(rxr_pkt_type_is_mulreq(pkt_type));
+	ep = ope->ep;
+	assert(ep->efa_max_outstanding_tx_ops >=
+	       ep->efa_outstanding_tx_ops + ep->efa_rnr_queued_pkt_cnt);
 
-	max_req_data_capacity = efa_rdm_txe_max_req_data_capacity(ep, txe, pkt_type);
-	assert(max_req_data_capacity);
+	available_tx_pkts = ep->efa_max_outstanding_tx_ops
+		- ep->efa_outstanding_tx_ops
+		- ep->efa_rnr_queued_pkt_cnt;
 
-	mulreq_total_data_size = efa_rdm_ope_mulreq_total_data_size(txe, pkt_type);
-	assert(mulreq_total_data_size);
+	if (available_tx_pkts == 0)
```

Comment:
How can we guarantee this number is non-negative. Will it be a negative number?

Reply:
The way it is calculated 

```
available_tx_pkts  = ep->efa_max_outstanding_tx_ops -
	       ep->efa_outstanding_tx_ops - ep->efa_rnr_queued_pkt_cnt);
```
Gurantee it will not be negative. because `ep->efa_max_outstanding_tx_ops` is total TX packet number. 
Each inflight send consume 1 packet (`ep->efa_outstanding_tx_ops`), and each queued RNR packet consume 1 packet.

There is an assertion that ensure `ep->efa_max_outstanding_tx_ops >= ep->efa_outstanding_tx_ops  + ep->efa_rnr_queued_pkt_cnt`-